include/vcpkg/base/util.h

```diff
@@ -60,6 +60,22 @@ namespace vcpkg::Util
         {
             return container.find(item) != container.end();
         }
+
+        template<class Container, class Pred>
+        void erase_if(Container& container, Pred pred)
```

Comment:
Given that `erase_if`'s whole reason to be in `std::` is to be container agnostic as far as the caller is concerned, I would like to see this moved out of sets somehow. For example, by SFINAE constraints:

```diff
diff --git a/include/vcpkg/base/util.h b/include/vcpkg/base/util.h
index 12207d99..3fec3889 100644
--- a/include/vcpkg/base/util.h
+++ b/include/vcpkg/base/util.h
@@ -60,20 +60,28 @@ namespace vcpkg::Util
         {
             return container.find(item) != container.end();
         }
+    }
+
+    template<class Container, class = void>
+    constexpr bool container_set_erasable = false;
+
+    template<class Container>
+    constexpr bool container_set_erasable<Container, std::void_t<decltype(
+        std::declval<typename Container::iterator&>() = std::declval<Container&>().erase(std::declval<typename Container::iterator&>())
+        )>> = true;
 
-        template<class Container, class Pred>
-        void erase_if(Container& container, Pred pred)
+    template<class Container, class Pred, std::enable_if_t<container_set_erasable<Container>, int> = 0>
+    void erase_if(Container& container, Pred pred)
+    {
+        for (auto i = container.begin(), last = container.end(); i != last;)
         {
-            for (auto i = container.begin(), last = container.end(); i != last;)
+            if (pred(*i))
             {
-                if (pred(*i))
-                {
-                    i = container.erase(i);
-                }
-                else
-                {
-                    ++i;
-                }
+                i = container.erase(i);
+            }
+            else
+            {
+                ++i;
             }
         }
     }
diff --git a/src/vcpkg/commands.ci.cpp b/src/vcpkg/commands.ci.cpp
index fd128c24..7e9c2bb9 100644
--- a/src/vcpkg/commands.ci.cpp
+++ b/src/vcpkg/commands.ci.cpp
@@ -500,8 +500,8 @@ namespace vcpkg::Commands::CI
         {
             StatusParagraphs status_db = database_load_check(paths.get_filesystem(), paths.installed());
             auto already_installed = SetInstalled::adjust_action_plan_to_status_db(action_plan, status_db);
-            Util::Sets::erase_if(already_installed,
-                                 [&](auto& spec) { return Util::Sets::contains(split_specs->known, spec); });
+            Util::erase_if(already_installed,
+                          [&](auto& spec) { return Util::Sets::contains(split_specs->known, spec); });
             if (!already_installed.empty())
             {
                 msg::println_warning(msgCISkipInstallation, msg::list = Strings::join(", ", already_installed));
```

or by just naming the type explicitly:

```diff
diff --git a/include/vcpkg/base/util.h b/include/vcpkg/base/util.h
index 12207d99..3e8b34f5 100644
--- a/include/vcpkg/base/util.h
+++ b/include/vcpkg/base/util.h
@@ -6,6 +6,7 @@
 #include <algorithm>
 #include <functional>
 #include <map>
+#include <set>
 #include <type_traits>
 #include <unordered_map>
 #include <utility>
@@ -60,20 +61,20 @@ namespace vcpkg::Util
         {
             return container.find(item) != container.end();
         }
+    }
 
-        template<class Container, class Pred>
-        void erase_if(Container& container, Pred pred)
+    template<class Pred, class... SetArgs>
+    void erase_if(std::set<SetArgs...>& container, Pred pred)
+    {
+        for (auto i = container.begin(), last = container.end(); i != last;)
         {
-            for (auto i = container.begin(), last = container.end(); i != last;)
+            if (pred(*i))
             {
-                if (pred(*i))
-                {
-                    i = container.erase(i);
-                }
-                else
-                {
-                    ++i;
-                }
+                i = container.erase(i);
+            }
+            else
+            {
+                ++i;
             }
         }
     }
diff --git a/src/vcpkg/commands.ci.cpp b/src/vcpkg/commands.ci.cpp
index fd128c24..7e9c2bb9 100644
--- a/src/vcpkg/commands.ci.cpp
+++ b/src/vcpkg/commands.ci.cpp
@@ -500,8 +500,8 @@ namespace vcpkg::Commands::CI
         {
             StatusParagraphs status_db = database_load_check(paths.get_filesystem(), paths.installed());
             auto already_installed = SetInstalled::adjust_action_plan_to_status_db(action_plan, status_db);
-            Util::Sets::erase_if(already_installed,
-                                 [&](auto& spec) { return Util::Sets::contains(split_specs->known, spec); });
+            Util::erase_if(already_installed,
+                          [&](auto& spec) { return Util::Sets::contains(split_specs->known, spec); });
             if (!already_installed.empty())
             {
                 msg::println_warning(msgCISkipInstallation, msg::list = Strings::join(", ", already_installed));
```

Reply:
Why SFINE at all? We could simply accept a `Container` and maybe add an overload for vectors to do the same as `erase_remove_if`

Reply:
That's the 'just naming the type explicitly' solution.